cpp/merkletree/sparse_merkle_tree.h

```diff
@@ -0,0 +1,265 @@
+
+  // Get the current root of the tree.
+  // Update the root to reflect the current shape of the tree,
+  // and return the tree digest.
+  //
+  // Returns the hash of an empty string if the tree has no leaves
+  // (and hence, no root).
+  virtual std::string CurrentRoot();
+
+  // Get the Merkle path from the leaf at |path| to the current root.
+  //
+  // Returns a vector of node hashes, ordered by levels from leaf to root.
+  // The first element is the sibling of the leaf hash, and the last element
+  // is one below the root.
+  // Returns an empty vector if the tree is not large enough
+  // or the leaf index is 0.
+  //
+  // @param path the path of the leaf whose inclusion proof to return.
+  std::vector<std::string> InclusionProof(const Path& path);
+
+  std::string Dump() const;
+
+ private:
+  // WARNING WARNING WARNING
+  // 64 < 256 !
+  // WARNING WARNING WARNING
+  // TODO(alcutter): BIGNUM probably.
+  typedef uint64_t IndexType;
+
+  struct Leaf {
+    Leaf(const Path& path, const std::string& value)
+        : path_(path), value_(value) {
+    }
+
+    std::string DebugString() const;
+
+    const Path path_;
+    std::string value_;
+  };
+
+  struct TreeNode {
+    TreeNode() : type_(INTERNAL), hash_(nullptr) {
+    }
+    TreeNode(const std::string& hash)
+        : type_(INTERNAL), hash_(new std::string(hash)) {
+    }
+
+    TreeNode(const Path& path, const std::string& leaf_value)
+        : type_(LEAF), leaf_(new Leaf(path, leaf_value)) {
+    }
+
+    std::string DebugString() const;
+
+    enum { INTERNAL, LEAF } type_;
+    // TODO(alcutter): sort this out
+    std::unique_ptr<std::string> hash_;
+    std::unique_ptr<Leaf> leaf_;
+  };
+
+  std::string CalculateSubtreeHash(size_t depth, IndexType index);
+
+  void DumpTree(std::ostream* os, size_t depth, IndexType index) const;
+
+  // Get the |index|-th node at level |level|. Indexing starts at 0;
+  // caller is responsible for ensuring tree is sufficiently up to date.
+  std::string Node(size_t level, size_t index) const;
+
+  // Maybe add a new tree level.
+  void EnsureHaveLevel(size_t n);
+
+  std::unique_ptr<SerialHasher> serial_hasher_;
+  TreeHasher treehasher_;
+  const std::vector<std::string>& null_hashes_;
+  // TODO(alcutter): investigate other structures
+  std::vector<std::unordered_map<IndexType, TreeNode>> tree_;
+  std::string root_hash_;
+};
+
+
+// Pretty print a Path
+std::ostream& operator<<(std::ostream& out,
+                         const SparseMerkleTree::Path& path);
+
+
+// Creates a Path from the bits passed in.
+inline SparseMerkleTree::Path PathFromBytes(const std::string& bytes) {
+  SparseMerkleTree::Path path;
+  // Path size must be a multiple of 8 for now.
+  CHECK_EQ(bytes.size(), path.size());
+
+  for (int i(0); i < path.size(); ++i) {
+    path[i] = bytes[i];
+  }
+  return path;
+}
+
+
+// Extracts the |n|th bit from |path|
+inline int PathBit(const SparseMerkleTree::Path& path, size_t bit) {
```

Comment:
But then I'd have to add the ternary operator in (currently) 3 extra places in `sparse_merkle_tree.cc` instead...